### ISSUE TITLE
[Link Handling] Make link references consistent and flexible

### ISSUE THREAD

**Issue by stefanseifert** [adobe/aem-core-wcm-components#713]
## Feature Request

(this is a follow-up from [this dicussion](https://groups.google.com/forum/#!msg/aem-core-components-dev/st_31IWmyhs/fT5ZCXT3DAAJ) in the mailing list)

currently, the storing of link reference data in properties and providing access to link references via the model interfaces is very inconsistent, see [1] for an overview. additionally, it's inflexible as it only allows to store a link url, and no further properties e.g. if the link should be opened in a new tab/window (_blank) or not. PR #625 is an attempt to add this to some components, but in my pov this is not flexible enough.

my proposal is:

* a) introduce an interface for link target reference
* b) all model interface that have a link (including ListItem) extend it
* c) include further properties that may come with the link like: target, SEO, tracking properties in a generic way
* d) change the HTL scripts to generate links always based in the methods of this new interface
* e) create a shared piece of TouchUI dialog definition (e.g. a field set) that is used in all edit dialogs to edit the target link and always looks the same

such an interface may look like this:

```java
interface LinkMixin {
  boolean isLinkValid();
  String getLinkURL();
  Map<String, String> getLinkAttributes();
}
```

the map returns key/value pairs like target=_blank, rel=nofollow etc. it should also contain the main href=xxx attribute. the separate getLinkURL method is only for components that for some reason do not generate an anchor tag but want to do other things with the LinkURL (e.g. putting it in a data attribute).

the HTL would look like:

```html
<a data-sly-unwrap="${!obj.linkValid }" data-sly-attribute="${obj.linkAttributes}">...</a>
```

the inconsistent property names cannot be cleaned up without breaking existing content, so we stick with them. for the interface which use link reference method not in compliance with the new method names a default implementation can be added that calls the new names.

[1] Components and their link properties/methods in Core Components 2.5.0

```
Component     | Property Name  | API Method
--------------+----------------+------------
Button        | link           | getLink
Image         | linkURL        | getLink
Teaser        | linkURL        | getLinkURL
Teaser/Action | link           | getURL
Title         | linkURL        | getLinkURL
ListItem      |                | getURL
```

**Comment by kwin** [comment]
Instead of relying on the HTML specific anchor attributes it would be nice to have some kind of abstraction for common attributes to be completely view agnostic and keep the model/view separation (i.e. dedicated getters per aspect). The problem is though that the list is probably not finite and may grow over time.

**Comment by stefanseifert** [comment]
another problem when using dedicated getters: if you add a new one, you have to change all markup templates of all components (probably including those customized by the downstream projects). i wanted to avoid this.

maybe we should rename `getLinkAttributes` to `getLinkHtmlAttributes` to make it clear it is not view agnostic?

**Comment by stefanseifert** [comment]
i've provided a PR for this issue #726 

this PR covers all changes in the Model APIs and the model implementations and the HTL markup files.
it should not change any behavior and should be fully backwards compatibly.

for supporting to set a window target for all editable links i've created a separate issue #725 to not bloat this issue too much.

**Comment by kwin** [comment]
As already mentioned in https://github.com/adobe/aem-core-wcm-components/pull/862#issuecomment-591325034 I think we need to clarify whether the url returns the mapped or unmapped url. As both is useful I would propose to add an additional `getUnmappedUrl()` which can be used whenever the AEM Link Rewriter is active, for all other cases rather `getUrl()` should be used, which should return the mapped URL (for JSON or in general in all contexts where the Link Rewriter is inactive)

**Comment by vladbailescu** [comment]
We decided to take a gradual approach to merging the proposed changes. We are breaking down @stefanseifert 's contribution (which was already merged to `3.0.0` branch) and bringing the changes in gradually:
1. Take in the new `Link` interface and the helpers (https://github.com/adobe/aem-core-wcm-components/pull/1411)
2. Adjust existing components and make sure all link processing is done with the newly-introduced helpers (this is done by @stefanseifert for most of the components, we will need to adapt changes to the refactoring that happened meanwhile and to make sure any new introduced functionality uses the helpers as well).
3. Stage new versions of the components that use links in separate feature branches, apply @stefanseifert's changes regarding link handling and review open issues (especially those that would require incompatible changes) to make sure all of them are applied before we merge the new version.

We believe this will helps us move further and avoid the situation that kept repeating in the past where changes and refactoring kept making merging this change harder.

**Comment by vladbailescu** [comment]
We will be releasing the new `Link` interface and `LinkHandler` helper as part of 2.16.0, with updated existing component versions. For each component that required new versions to take better advantage of the functionality changes we staged the changes in a feature branch. We will pick these later and make sure to also fix other issues that would require non-backwards-compatible changes at the same time: https://github.com/adobe/aem-core-wcm-components/pulls?q=is%3Apr+LinkHandler+Stage3
